### PR TITLE
Irm preliminary data

### PR DESCRIPTION
**Pull request by ben-heil**
This PR runs a preliminary proof-of-concept analysis for using invariant risk minimization on gene expression data.

The main changes are:
1. Creating the `IRM_analysis` notebook
2. Avoiding adding directories to the pythonpath manually by finally making the repo into a python package (modifies `setup.py`, `__init__.py`, and the first few lines of the modules) 
3. Change the manual logging of results to use a tensorboard writer (modifies `train.py`
4. Create a function to calculate weights to address the class imbalance in sepsis case/control statis (modifies `utils.py`)

I'm sorry for the messiness of this PR, I started working on the notebook then found other things I had to change to make it work the way I wanted it to.

### CODE REVIEW

**Comment by danich1** [comment]
Can you convert the IRM_analysis into a script? (`jupyter nbconvert irm_analysis.ipynb --to script` ) Would make review a bit easier.

**Comment by ben-heil** [comment]
Sorry, just now saw your comment. There is now a converted version of the notebook.

File: environment.yml [hunk 1]
```diff
@@ -3,133 +3,169 @@ channels:
   - conda-forge
   - defaults
 dependencies:
```
Comment (danich1) [inline environment.yml:5]: Did you generate this with `conda list`? If so that's fine, but I would recommend keeping this file to only contain packages that need to be installed.  (Personal preference so you can ignore if you think otherwise)

File: notebook/converted/IRM_analysis.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+# In[7]:
+
+
+sepsis_metadata_df['platform'].value_counts()
+
+
+# In[8]:
+
+
+sepsis_metadata_df[sepsis_metadata_df['platform'] == 'affymetrix human genome u133a array (hgu133a)']
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:103]: Tip: you can use pandas query:
```python
sepsis_metadata_df.query(
"platform == 'affymetrix human genome u133a array (hgu133a)'"
)
```
supports f-strings as well

File: notebook/converted/IRM_analysis.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+# ## IRM Evaluation
+
+# ### Setup
+# This code is unnecessary, as the variables are already initialized in the EDA section.
+# It is included here for the sake of clarity for people who skip to the analysis
+
+# In[11]:
+
+
+curr_path = os.path.dirname(os.path.abspath(os.path.abspath('')))
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:130]: Tip:
```python
from pathlib import Path
str(Path("./data/sample_classification.pkl"))
```
pathlib is pretty powerful when forming directory paths. Plus it can support f-strings as well.

File: notebook/converted/IRM_analysis.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+# In[13]:
+
+
+tune_df = sepsis_metadata_df[sepsis_metadata_df['platform'] == 'affymetrix human genome u219 array (hgu219)']
+train_df = sepsis_metadata_df[sepsis_metadata_df['platform'] != 'affymetrix human genome u219 array (hgu219)']
+print(len(tune_df.index))
+print(len(train_df.index))
+
+tune_studies = tune_df['study'].unique()
+tune_dataset = CompendiumDataset(tune_studies, classes, sample_to_label, metadata_path, compendium_path, encoder)
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:163]: Will be a nitpicker here: add white space for long arguments
```python
tune_dataset = CompendiumDataset(
  tune_studies, classes, 
  sample_to_label, metadata_path, 
  compendium_path, encoder
)
```

File: notebook/converted/IRM_analysis.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+
+# ### Filter Platforms
+# Remove a platform that corresponds to a study present in the labeled data, but not the human compendium
+
+# In[14]:
+
+
+platforms = train_df['platform'].unique()
+platforms = [p for p in platforms if p != 'affymetrix human human exon 1.0 st array (huex10st)']
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:174]: Same as above whitspace:
```python
platforms = [
  p 
  for p in platforms
  if p!= 'affymetrix human human exon 1.0 st array (huex10st)'
]
```
Just makes things easier to process for other people reading your code.

File: notebook/converted/IRM_analysis.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    full_train_loader = DataLoader(full_train_dataset, batch_size=8, shuffle=True)
+
+    assert total_irm_samples == len(full_train_dataset)
+    
+    for seed in range(num_seeds):
+        print(hold_out_platform, seed)
+        np.random.seed(seed)
+        random.seed(seed)
+        net = models.ThreeLayerNet(len(compendium_df.index))
+        writer = SummaryWriter(os.path.join(curr_path,
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:215]: Refer to pathlib/white space comment.

File: notebook/converted/IRM_analysis.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        results = train.train_with_erm(net, full_train_loader, tune_loader, num_epochs=400, 
+                                       save_file=save_file, writer=writer)
+        erm_result_list.append(results)
+
+        net = models.ThreeLayerNet(len(compendium_df.index))
+        writer = SummaryWriter(os.path.join(curr_path, os.pardir, 'logs', 
+                                            'irm_analysis_{}_{}.tfrecord'.format(platform_file, seed)))
+
+        save_file = os.path.join(curr_path, os.pardir, 'logs', 'irm_analysis_{}_{}.pkl'.format(platform_file, seed))
+        results = train.train_with_irm(net, train_loaders, tune_loader, num_epochs=400,
```
Comment (danich1) [inline notebook/converted/IRM_analysis.py:230]: White Space comment above

File: whistl/datasets.py [hunk 8]
```diff
@@ -442,16 +431,84 @@ def __len__(self):
...
+        try:
+            experiment_samples = experiment_metadata[experiment]['sample_accession_codes']
+            sample_list.extend(experiment_samples)
+        except KeyError as e:
+            # If an experiment doesn't have any samples for some reason, skip it
+            print(e)
+            pass
+
+
+    for sample in sample_list:
```
Comment (danich1) [inline whistl/datasets.py:463]: Tip:

```python
sample_to_platform = {
sample:sample_metadata[sample]
for sample in sample list
}
```
^ Can turn this into a dictionary comprehension since sample_to_platform is only used one.

File: whistl/datasets.py [hunk 9]
```diff
@@ -442,16 +431,84 @@ def __len__(self):
...
+    subset_df = expression_df.loc[:, samples_to_keep]
+
+    return subset_df
+
+
 class CompendiumDataset(Dataset):
     '''A dataset of one or more studies pulled from the refine.bio human compendium'''
 
-    def __init__(self, studies, classes, sample_to_label, metadata_path, compendium_path, encoder):
+    def __init__(self, groups, classes, sample_to_label, metadata_path, compendium_path, encoder,
```
Comment (danich1) [inline whistl/datasets.py]: It might be my github but why is there a new line near the last argument? Also consider adding whitespace to header since the number of arguments have grown.

**Review by danich1 — COMMENTED** [review]
Looking pretty good. Main gripe is white space in code. Blame my comp sci teachers in undergrad, but they were right that it makes code SO much easier to read. 

Few questions:
1. erm stands for?
2. In the figure 'F1 Score on held-out platform' what is going on with that outlier on the irm column? Also is that particular platform category harder to predict under the IRM assumption?
3. Same questions as above for the 'Mean accuracies in IRM and ERM on tune set' plot.
4. Did you mean to plot the same figures twice in your notebook or is the last two plots suppose to be the test set?

Will approve when questions are answered and my small nitpicking has been addressed!

File: environment.yml (hunk 1 above)
Comment (ben-heil) [reply environment.yml:5]: I use `conda env export --no-builds`. It would be more readable with just the main packages, but including all the dependencies used by libraries in the whistl environment ensures identical behavior between machines

File: notebook/converted/IRM_analysis.py (hunk 2 above)
Comment (ben-heil) [reply notebook/converted/IRM_analysis.py:103]: Huh, I didn't know about that. That's super useful, thanks!

File: notebook/converted/IRM_analysis.py (hunk 5 above)
Comment (ben-heil) [reply notebook/converted/IRM_analysis.py:174]: That's cool, I didn't know you could to a list comprehension across lines like that

File: whistl/datasets.py (hunk 8 above)
Comment (ben-heil) [reply whistl/datasets.py:463]: That's a good idea. In this case it would either push me past my line limit, or make me change my indentation style, but I will look for the pattern in the future

File: whistl/datasets.py (hunk 9 above)
Comment (ben-heil) [reply whistl/datasets.py]: The newline was to keep the function definition from exceeding the line length limit. I added more whitespace so it's no longer an issue

**Comment by ben-heil** [comment]
> Looking pretty good. Main gripe is white space in code. Blame my comp sci teachers in undergrad, but they were right that it makes code SO much easier to read.
> 
> Few questions:
> 
> 1. erm stands for?
> 2. In the figure 'F1 Score on held-out platform' what is going on with that outlier on the irm column? Also is that particular platform category harder to predict under the IRM assumption?
> 3. Same questions as above for the 'Mean accuracies in IRM and ERM on tune set' plot.
> 4. Did you mean to plot the same figures twice in your notebook or is the last two plots suppose to be the test set?
> 
> Will approve when questions are answered and my small nitpicking has been addressed!

1. ERM stands for "Empirical risk minimization", which is the IRM paper's authors' way of framing standard network training via SGD

2. IRM is fairly unstable, which may be part of it. If I had to speculate, I would guess that the data for that platform is dissimilar to the other platforms, so it is more difficult to predict, especially given that the best performance on the tuning set comes when it is left out. That theory isn't consistent with it being predicted alright when trained with ERM though. If you're talking about the single outlier point, I think that is just instability in training, you can see a pretty big difference between the median and the mean accuracy for that platform in the later figures.

I realized when addressing your comments that I only set the random seed for numpy and python, not torch, so I guess we'll see whether another draw from the same distribution shows the same pattern.

3. The accuracies on the tune set are easier to answer. The u133 plus 2.0 array accounts for ~3/4 of the data in the sepsis dataset, so dropping it reduces the accuracy more than losing any other study.

4. You're right that this was unclear, I added a few markdown cells to explain. The last two figures measure the median instead of the mean because the IRM training curves mostly converge to a single tune set accuracy, but will have large loss/accuracy spikes occasionally during training. As a result, measuring the median accuracy may make more sense for estimating central tendency than measuring the mean

**Review by danich1 — APPROVED** [review]
LGTM! Interesting that IRM can be unstable during training. Looking forward to see how this evolves.
